framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -3106,7 +3106,7 @@ public boolean validateTypeOf(Tree tree) {
             }
             break;
         default:
-            type = atypeFactory.getAnnotatedType(tree);
+        	type = atypeFactory.getDefaultedAnnotatedType(tree);
```

Comment:
I imagined a system in which C and D would be valid.
I would only expect an error for the B that is introduced as LUB of the conditional expression; I don't want an error for the type of `cond`.
I prefer checking all types for validity, even though this might be redundant. At some point we can have a look and see whether it's worth trying to optimize this to only check explicit types.
I'll try reproducing this error in a standard type system and will get back to you.